File: src/main/java/com/back/mozu/domain/setting/service/HolidayService.java
package com.back.mozu.domain.setting.service;

import com.back.mozu.domain.reservation.repository.ReservationRepository;
import com.back.mozu.domain.setting.dto.HolidayDto;
import com.back.mozu.domain.setting.entity.Holiday;
import com.back.mozu.domain.setting.repository.HolidayRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.LocalDate;
import java.time.YearMonth;
import java.util.List;

@Service
@RequiredArgsConstructor
@Transactional(readOnly = true)
public class HolidayService {

    private final HolidayRepository holidayRepository;
    private final ReservationRepository reservationRepository;

    public HolidayDto.GetHolidaysResponse getHolidays(String month) {
        List<Holiday> holidays;

        if (month == null || month.isBlank()) {
            holidays = holidayRepository.findAll()
                    .stream()
                    .sorted((a, b) -> a.getDate().compareTo(b.getDate()))
                    .toList();
        } else {
            YearMonth yearMonth = YearMonth.parse(month);
            LocalDate startDate = yearMonth.atDay(1);
            LocalDate endDate = yearMonth.atEndOfMonth();
            holidays = holidayRepository.findByDateBetweenOrderByDateAsc(startDate, endDate);
        }

        List<HolidayDto.HolidayItem> items = holidays.stream()
                .map(HolidayDto.HolidayItem::from)
                .toList();

        return new HolidayDto.GetHolidaysResponse(items.size(), items);
    }

    @Transactional
    public HolidayDto.CreateHolidayResponse createHoliday(HolidayDto.CreateHolidayRequest request) {
        validateCreateRequest(request);

        Holiday holiday = Holiday.builder()
                .date(request.date())
                .reason(request.reason())
                .build();

        Holiday savedHoliday = holidayRepository.save(holiday);
        int conflictingReservationCount = reservationRepository.countByTimeSlot_Date(savedHoliday.getDate());

        return new HolidayDto.CreateHolidayResponse(
                savedHoliday.getDate(),
                savedHoliday.getReason(),
                conflictingReservationCount
        );
    }

    private void validateCreateRequest(HolidayDto.CreateHolidayRequest request) {
        if (request.date().isBefore(LocalDate.now())) {
            throw new IllegalArgumentException("과거 날짜는 휴무일로 설정할 수 없습니다.");
        }

        if (holidayRepository.existsById(request.date())) {
            throw new IllegalStateException("이미 등록된 휴무일입니다.");
        }
    }
}

File: src/main/java/com/back/mozu/domain/setting/service/HolidayService.kt
package com.back.mozu.domain.setting.service

import com.back.mozu.domain.reservation.repository.ReservationRepository
import com.back.mozu.domain.setting.dto.HolidayDto.*
import com.back.mozu.domain.setting.entity.Holiday
import com.back.mozu.domain.setting.repository.HolidayRepository
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import java.time.LocalDate
import java.time.YearMonth

@Service
@Transactional(readOnly = true)
// @RequiredArgsConstructor 제거 → 코틀린 생성자 주입으로 대체
class HolidayService(
    private val holidayRepository: HolidayRepository,
    private val reservationRepository: ReservationRepository
) {

    fun getHolidays(month: String?): GetHolidaysResponse {
        // MutableList<Holiday?>? 제거 → 타입 추론으로 대체
        // if-else 표현식으로 변환 (Java의 if문 → Kotlin의 if 표현식)
        val holidays = if (month == null || month.isBlank()) {
            // stream().sorted() → sortedBy()로 대체
            holidayRepository.findAll().sortedBy { it.date }
        } else {
            val yearMonth = YearMonth.parse(month)
            holidayRepository.findByDateBetweenOrderByDateAsc(
                yearMonth.atDay(1),
                yearMonth.atEndOfMonth()
            )
        }
        // stream().map().toList() → map {}으로 대체
        val items = holidays.map { HolidayItem.from(it) }

        return GetHolidaysResponse(items.size, items)
    }

    @Transactional
    fun createHoliday(request: CreateHolidayRequest): CreateHolidayResponse {
        validateCreateRequest(request)

        // Builder 패턴 제거 → named argument로 대체
        val savedHoliday = holidayRepository.save(
            Holiday(date = request.date, reason = request.reason)
        )
        // !! 제거 → 생성자 주입으로 non-null 보장
        val conflictingReservationCount = reservationRepository.countByTimeSlot_Date(savedHoliday.date)

        return CreateHolidayResponse(
            savedHoliday.date,
            savedHoliday.reason,
            conflictingReservationCount
        )
    }

    private fun validateCreateRequest(request: CreateHolidayRequest) {
        require(!request.date.isBefore(LocalDate.now())) { "과거 날짜는 휴무일로 설정할 수 없습니다." }
        check(!holidayRepository.existsById(request.date)) { "이미 등록된 휴무일입니다." }
    }
}